kythe/go/platform/tools/kzip/stripcmd/stripcmd.go

```diff
@@ -0,0 +1,152 @@
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"path/filepath"
+
+	"kythe.io/kythe/go/platform/kzip"
+	"kythe.io/kythe/go/platform/tools/kzip/flags"
+	"kythe.io/kythe/go/platform/vfs"
+	"kythe.io/kythe/go/util/cmdutil"
+
+	"bitbucket.org/creachadair/stringset"
+	"github.com/google/subcommands"
+)
+
+type stripCommand struct {
+	cmdutil.Info
+
+	input    string
+	output   string
+	append   bool
+	encoding flags.EncodingFlag
+}
+
+// New creates a new subcommand for merging kzip files.
+func New() subcommands.Command {
+	return &stripCommand{
+		Info:     cmdutil.NewInfo("strip", "strip units from kzip file", "--input path --output path unit-hash*"),
+		encoding: flags.EncodingFlag{Encoding: kzip.EncodingJSON},
+	}
+}
+
+// SetFlags implements the subcommands interface and provides command-specific flags
+// for merging kzip files.
+func (c *stripCommand) SetFlags(fs *flag.FlagSet) {
+	fs.StringVar(&c.output, "output", "", "Path to output kzip file")
+	fs.StringVar(&c.input, "input", "", "Path to input kzip file")
+	fs.Var(&c.encoding, "encoding", "Encoding to use on output, one of JSON, PROTO, or ALL")
+}
+
+// Execute implements the subcommands interface and merges the provided files.
+func (c *stripCommand) Execute(ctx context.Context, fs *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	if c.output == "" {
+		return c.Fail("required --output path missing")
+	}
+	if c.input == "" {
+		return c.Fail("required --input path missing")
+	}
+	opt := kzip.WithEncoding(c.encoding.Encoding)
+	dir, file := filepath.Split(c.output)
+	if dir == "" {
+		dir = "."
+	}
+	tmpOut, err := vfs.CreateTempFile(ctx, dir, file)
+	tmpName := tmpOut.Name()
+	defer func() {
+		if tmpOut != nil {
+			tmpOut.Close()
+			vfs.Remove(ctx, tmpName)
+		}
+	}()
+	if err != nil {
+		return c.Fail("Error creating temp output: %v", err)
+	}
+	units := stringset.FromKeys(fs.Args())
+	if err := stripArchive(ctx, tmpOut, c.input, units, opt); err != nil {
+		return c.Fail("Error stripping archives: %v", err)
+	}
+	if err := vfs.Rename(ctx, tmpName, c.output); err != nil {
+		return c.Fail("Error renaming tmp to output: %v", err)
+	}
+	return subcommands.ExitSuccess
+}
+
+func stripArchive(ctx context.Context, out io.WriteCloser, input string, digests stringset.Set, opts ...kzip.WriterOption) error {
+	wr, err := kzip.NewWriteCloser(out, opts...)
+	if err != nil {
+		out.Close()
+		return fmt.Errorf("error creating writer: %v", err)
+	}
+
+	filesAdded := stringset.New()
+
+	f, err := vfs.Open(ctx, input)
+	if err != nil {
+		return fmt.Errorf("error opening archive: %v", err)
+	}
+	defer f.Close()
+
+	stat, err := vfs.Stat(ctx, input)
+	if err != nil {
+		return err
+	}
+	size := stat.Size()
+	if size == 0 {
+		log.Printf("Skipping empty .kzip: %s", input)
+		return nil
+	}
+
+	rd, err := kzip.NewReader(f, size)
+	if err != nil {
+		return fmt.Errorf("error creating reader: %v", err)
+	}
+
+	err = rd.Scan(func(u *kzip.Unit) error {
+		if !digests.Contains(u.Digest) {
+			return nil
```

Comment:
maybe I just stepped through this backwards, but does this mean that only units not in the commandline args will get included in the output?  Is that the opposite of what you want?

Reply:
It's correct. If the specified digests do not contain the unit being scanned, skip it.

Reply:
Should this be renamed `filter` instead of `strip`?

Reply:
that "strip" suggests removing (rather than keeping) the named CUs is my main concern

Reply:
renamed

Reply:
Ahh ok yea that was my confusion then.